nkms/storage/encrypted_file.py

```diff
@@ -0,0 +1,200 @@
+        An EncryptedFile object actually is composed of two files:
+        1) The ciphertext -- This is the chunked and encrypted ciphertext
+        2) The header -- This contains the metadata of the ciphertext that
+            tells us how to decrypt it, or add more data.
+
+        :param bytes key: Symmetric key to use for encryption/decryption
+        :param string/bytes path: Path of file to open
+        :param string mode: Mode to use when opening file, default is 'rb'
+        """
+        self.path = path
+        self.mode = mode
+
+        cipher = symmetric_from_algorithm(default_algorithm)
+        self.cipher = cipher(key)
+
+    def _build_header(self, version=100, nonce=None, keys=None,
+                      chunk_size=1000000, num_chunks=0, msg_len=0):
+        """
+        Builds a header and returns the msgpack encoded form of it.
+
+        :param int version: Version of the NuCypher header
+        :param bytes nonce: Nonce to write to header, default is random(20)
+        :param list keys: Keys to write to header
+        :param int chunk_size: Size of each chunk in bytes, default is 1MB
+        :param int num_chunks: Number of chunks in ciphertext, default is 0
+        :param int msg_len: Length of the encrypted ciphertext in total
+
+        :return: (header_length, encoded_header)
+        :rtype: Tuple(int, bytes)
+        """
+        if not nonce:
+            nonce = random(20)
+
+        self.header = {
+            'version': version,
+            'nonce': nonce,
+            'keys': keys,
+            'chunk_size': chunk_size,
+            'num_chunks': num_chunks,
+            'msg_len': msg_len,
+        }
+
+        try:
+            encoded_header = msgpack.dumps(self.header)
+        except ValueError as e:
+            raise e
+        self.header_length = len(encoded_header)
+        return (self.header_length, encoded_header)
+
+    def _encode_header(self):
+        """
+        Returns a msgpack encoded header and the length of it in bytes ready to
+        be written to the file_obj.
+
+        :return: (encoded_header, header_length_bytes)
+        :rtype: Tuple(bytes, bytes)
+        """
+        header_length_bytes = self.header_length.to_bytes(4, byteorder='big')
+        encoded_header = msgpack.dumps(self.header)
+        return (encoded_header, header_length_bytes)
+
+    def _update_header(self, header):
+        """
+        Updates the self.header with the key/values in header, then updates
+        the header length.
+
+        :param dict header: Dict to update self.header with
+
+        :return: (header_length, encoded_header)
+        :rtype: Tuple(int, bytes)
+        """
+        self.header.update(header)
+        try:
+            encoded_header = msgpack.dumps(self.header)
+        except ValueError as e:
+            raise e
+        self.header_length = len(encoded_header)
+        return (self.header_length, encoded_header)
+
+    def _read_header(self):
+        """
```

Comment:
Maybe it makes sense to turn 4 into a named constant, `HEADER_LENGTH`, for readability?